LDAP now also properly allows `nil` in names.

adam-james-v opened the pull request:
#fixes: #16158

This is a small edit to the LDAP implementation to allow `nil` in names, which is already the case for other SSO methods. An SSO refactor would be nice (and is in draft at https://github.com/metabase/metabase/pull/23651), but may take more time to do properly.

This PR makes the behaviour across the SSO methods more uniform.

adam-james-v committed:
LDAP now also properly allows `nil` in names.

There were still places where "Unknown" was substituted in to name keys, which we don't need to do anymore as 'nil' is
valid for first/last names of users.

adam-james-v committed:
Remove unused namespaces/refers

noahmoss commented on src/metabase/integrations/ldap/default_implementation.clj:
```diff
@@ -114,37 +113,26 @@
                          user-changes   (merge
-                                          (when-not (= new-first-name old-first-name) {:first_name new-first-name})
-                                          (when-not (= new-last-name old-last-name) {:last_name new-last-name}))]
+                                          (when (and first-name (not= first-name old-first-name)) {:first_name first-name})
+                                          (when (and last-name (not= last-name old-last-name)) {:last_name last-name}))]
```
Does it still make sense to avoid setting `nil` in the DB if there's already a value?

The original reasoning was if you don't have a name set in LDAP, it'll default to `Unknown` and then you can go in and edit it yourself. And we don't want to override that.

But you can't edit your name anymore if you've logged in with SSO, right? So maybe it would make sense to just set the LDAP value in all cases?

adam-james-v replied:
I thought so too, actually. There are a few tests that failed when I didn't have this change. https://github.com/metabase/metabase/blob/1aa9565c451fb2c92c4cab46c45dde7a0adab174/test/metabase/integrations/ldap_test.clj#L151-L157

Which imply a need to keep the app-db data around, although I think it's a fair question whether or not these tests are still relevant at all.

noahmoss replied:
Right, those are testing the old behavior. My question is about whether that behavior is still relevant. Probably a question for product.

adam-james-v committed:
Adjust LDAP so that even nil name values update the Metabase user

Also changed a test in both the OSS and EE LDAP tests to make sure this is indeed true that 'nil' values are correctly
set in the app-db when LDAP does not send `givenName` and/or `sn`.
